File: Day 2/solution.py
"""Checks what reports are safe and how many reports are safe with fault tolerance."""

def safe_report_count(row, min_val=1, max_val=3) -> int:
    """Counts the number of safe reports without fault tolerance."""
    return 1 if all(min_val <= row[i+1] - row[i] <= max_val for i in range(len(row) - 1)) \
             or all(min_val <= row[i] - row[i+1] <= max_val for i in range(len(row) - 1)) else 0

def safe_report_count_fault_tolerance(row) -> int:
    """Counts the number of safe reports with fault tolerance."""

    if safe_report_count(row):
        return 1

    safe: int = 0

    if safe_report_count(row):
        return 1

    for i in range(len(row)):
        fault_tolerated_report = row[:i] + row[i+1:]
        if safe_report_count(fault_tolerated_report):
            safe += 1
            break

    return safe

if __name__ == '__main__':
    safe_reports: int = 0
    safe_reports_fault_tolerance: int = 0

    with open('input.txt', "r", encoding="utf-8") as f:
        reports = f.readlines()

        for levels in reports:
            level_values = list(map(int, levels.split()))
            safe_reports += safe_report_count(level_values)
            safe_reports_fault_tolerance += safe_report_count_fault_tolerance(level_values)

    print(f"This is the safe amount of reports: {safe_reports}")
    print(f"This is the safe amount with one fault tolerance: {safe_reports_fault_tolerance}")

File: Day 2/solution.go
package main

import (
	"fmt"
	"os"
	"strings"
)

// Check if slice is made out of decreasing numbers.
func decreasing(row []int) bool {
	for i := 0; i < len(row)-1; i++ {
		diff := row[i+1] - row[i]
		if diff < 1 || diff > 3 {
			return false
		}
	}
	return true
}

// Check if slice is made out of increasing numbers.
func increasing(row []int) bool {
	for i := 0; i < len(row)-1; i++ {
		diff := row[i] - row[i+1]
		if diff < 1 || diff > 3 {
			return false
		}
	}
	return true
}

// Check if either condition are true.
func safeCheck(row []int) bool {
	return decreasing(row) || increasing(row)
}

// Transform String of seperated ints to type "int array/slice".
func stringToIntList(row string) []int {

	digitsStr := strings.Fields(row)
	digits := make([]int, len(digitsStr))
	for i, s := range digitsStr {
		fmt.Sscanf(s, "%d", &digits[i])
	}
	return digits
}

func safeReportCountFunc(row string) int {

	var safeStatus int

	digits := stringToIntList(row)
	if safeCheck(digits) {
		safeStatus++
	}

	return safeStatus

}

// Check if either condition are true, but with fault tolerance.
func safeReportCountFaultToleranceFunc(row string) int {

	digits := stringToIntList(row)

	if safeCheck(digits) {
		return 1
	}

	for i := 0; i < len(digits); i++ {

		temp := make([]int, len(digits))
		copy(temp, digits)

		faultToleratedReport := append(temp[:i], temp[i+1:]...)
		if safeCheck(faultToleratedReport) {
			return 1
		}
	}
	return 0
}

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func main() {

	var safeReportCount int
	var safeReportCountFaultTolerance int

	levels, err := os.ReadFile("input.txt")
	check(err)

	lines := strings.Split(string(levels), "\n")

	for _, row := range lines {
		safeReportCount += safeReportCountFunc(row)
		safeReportCountFaultTolerance += safeReportCountFaultToleranceFunc(row)
	}

	fmt.Println("Safe Report Count: ", safeReportCount)
	fmt.Println("Safe Report Count Fault Tolerance: ", safeReportCountFaultTolerance)

}
